pisound-btn/pisound-btn.c

```diff
@@ -231,96 +229,141 @@ static void read_config_value(const char *conf, const char *value_name, char *ds
-			*script = DEFAULT_HOLD_5S;
-		}
-		else
-		{
-			*name = HOLD_OTHER_VALUE_NAME;
-			*script = DEFAULT_HOLD_OTHER;
+			sprintf( ACTION_NAME, HOLD_NAME, timer );
+		} else {
+			strcpy( ACTION_NAME, HOLD_OTHER_VALUE_NAME );
 		}
 		break;
 	default:
-		*name = NULL;
-		*script = NULL;
 		break;
 	}
+	return ACTION_NAME[0] != '\0';
 }
 
-// Returns the length of the path or an error (negative number).
-static int get_action_script_path(enum action_e action, unsigned arg0, unsigned arg1, char *dst, char *args, size_t n)
+static bool get_action_script(enum action_e action, char* action_name, char *buffer, char* args, unsigned int buff_length)
 {
-	if (action < 0 || action >= A_COUNT)
-		return -EINVAL;
+	if (action_name[0] == '\0')
+	{
+		return false;
+	}
 
-	const char *action_name = NULL;
-	const char *default_script = NULL;
+	if (!read_config_value(g_config_path, action_name, buffer, args, buff_length, "\0")) // No value was specified.
+	{
+		bool found = false;
+		switch (action)
+		{
+		case A_CLICK:
+			if (strcmp("CLICK_1", action_name) == 0)
+			{
+				strcpy( buffer, DEFAULT_CLICK_1 );
+				if (args)
+					args[0] = '\0';
+				found = true;
+			}
+			else if (strcmp("CLICK_2", action_name) == 0)
+			{
+				strcpy( buffer, DEFAULT_CLICK_2 );
+				if (args)
+					args[0] = '\0';
+				found = true;
+			}
+			else if (strcmp("CLICK_3", action_name) == 0)
+			{
+				strcpy( buffer, DEFAULT_CLICK_3 );
+				if (args)
+					args[0] = '\0';
+				found = true;
+			}
+			else
+			{
+				found = read_config_value(g_config_path, CLICK_OTHER_VALUE_NAME, buffer, args, buff_length, "\0");
+			}
+			break;
+		case A_HOLD:
+			if (strcmp("HOLD_3S", action_name) == 0)
+			{
+				strcpy( buffer, DEFAULT_HOLD_3S );
+				if (args)
+					args[0] = '\0';
+				found = true;
+			}
+			else if (strcmp("HOLD_5S", action_name) == 0)
+			{
+				strcpy( buffer, DEFAULT_HOLD_5S );
+				if (args)
+					args[0] = '\0';
+				found = true;
+			} else {
+				found = read_config_value(g_config_path, HOLD_OTHER_VALUE_NAME, buffer, args, buff_length, "\0");
+			}
+			break;
+		default:
+			break;
+		}
+		return found;
+	}
+	return true;
+}
 
-	get_default_action_and_script(action, arg0, arg1, &action_name, &default_script);
+// Returns the length of the path or an error (negative number).
+static int get_action_script_path(enum action_e action, unsigned click_count, unsigned hold_time, char *script, char *args, size_t n)
+{
+	if (action < 0 || action >= A_COUNT) {
+		return -EINVAL;
+	}
 
-	if (action_name == NULL || default_script == NULL)
+	if (!get_action_name(action, click_count, hold_time))
+	{
 		return -EINVAL;
+	}
 
-	char script[MAX_PATH_LENGTH + 1];
+	char action_script[MAX_PATH_LENGTH + 1];
 	char arguments[MAX_PATH_LENGTH + 1];
 
-	read_config_value(g_config_path, action_name, script, arguments, MAX_PATH_LENGTH, default_script);
+	if (!get_action_script(action, ACTION_NAME, action_script, arguments, MAX_PATH_LENGTH ))
+	{
+		return 0;
```

Comment:
0 should mean success, if get_action_script was returning an errno int itself, you should just return it here too. Otherwise, some relevant negative number for the actual fault from errno.h could be used.